File: (4)/code/main.py
import numpy as np
from scipy.sparse import coo_matrix
from scipy.linalg import eigh
import matplotlib.pyplot as plt
import networkx as nx
from classes import Graph
from sklearn.cluster import KMeans

# Create a graph object
G = nx.read_edgelist('data/example1.dat', delimiter=',')
nx.draw(G, node_size=15)
plt.show()

def spectral_clustering(G):
    A = nx.adjacency_matrix(G)
    D = np.diagflat(np.sum(A, axis=1))
    D_inv = np.linalg.inv(D)
    L = np.dot(np.dot(D_inv, A), D_inv)

    eigenvalues, v = eigh(L)

    # Get the k largest eigenvectors
    k = np.ediff1d(eigenvalues).argmax() + 1
    print(k)
    X = v[:, -k:]
    Y = X / np.linalg.norm(X, axis=1, keepdims=True)
    kmeans = KMeans(n_clusters=k, random_state=0).fit(Y)

    _, fiedler_vector = eigh(D-A)
    fiedler_vector = fiedler_vector[:, 1]

    return kmeans.labels_, fiedler_vector, A

#labels, fiedler_vector, A = spectral_clustering(G)
G = Graph()

G(filename='data/example1.dat')

#G.visualize_graph()
#
G.number_of_clusters = 4

# Step 1. Construct the affinity matrix
#G.max_ids = np.max(G.nodes)
#G.A = np.zeros((G.max_ids, G.max_ids))
print(len(G.edges))
print(G.number_of_nodes())
print(G.number_of_edges())
for x in G.edges:
    print(x)
gamma = 1 # some constant - TODO: update this later!
for i, edge in enumerate(G.edges): # Iterate over the edges
    for j in range (len(G.edges)): #

        pairwise_similarity = np.exp((-abs(G[i] - G[j]))**2) / (2 * gamma**2)
        G.A[i][j] = pairwise_similarity
    
    #G.A[edge[0]-1, edge[1]-1] = 1

# Create the affinity matrix. Holds the pairwise similarity between nodes using a Gaussian kernel.
gamma = 1 # some constant - update this later!
#for np.exp(-abs(G[i] -))

#affinity_matrix = np.exp(-gamma * G.A)
#affinity_matrix = np.exp(-gamma * np.square(np.linalg.norm(G[:, None] - G, axis=2)))

# Construct the diagonal matrix
G.D = np.diag(np.sum(G.A, axis=1))
D_inv = np.linalg.inv(G.D)
G.L = np.dot(np.dot(D_inv, G.A), D_inv)

# Plot the eigenvalues
G.eigenvalues, v = eigh(G.L)
plt.plot(G.eigenvalues)
plt.title('Eigenvalues')
#plt.show()

# Get the k largest eigenvectors
# Sort the eigenvalues in ascending order
idx = G.eigenvalues.argsort()
k = G.number_of_clusters
X = v[:, idx[0:k]]
print(X.shape)

# Renormalize the eigenvectors to have unit length 
Y = np.zeros(X.shape)
for i in range(X.shape[1]):
    Y[:, i] = X[:, i] / np.linalg.norm(X[:, i])
    

# Cluster the data using k-means
from sklearn.cluster import KMeans
kmeans = KMeans(n_clusters=k, random_state=0).fit(Y)
G.cluster_labels = kmeans.labels_

# Visualize the graph
G.visualize_graph()

"""
It creates a sparse matrix As with dimensions max_ids x max_ids.
 The matrix is populated with 1’s at the locations specified by the vectors col1 and col2. 
 The remaining elements of the matrix are 0’s.
"""

File: (4)/code/main.m
%% Read data
clear all
k_1 = 4;
k_2 = 2;
E_1 = csvread('example1.dat');
E_2 = csvread('example2.dat');

%% Step 1
A_1 = step1(E_1);
A_2 = step1(E_2);
% Plot
figure(1);
subplot(1,2,1);
spy(A_1);
title("SparsityPattern for A_1");
subplot(1,2,2);
spy(A_2);
title("SparsityPattern for A_2");
%% Display Graph
figure(2);
subplot(1,2,1);
plot(graph(A_1), 'Layout','force');
title("Graph for A_1");
subplot(1,2,2);
plot(graph(A_2),'Layout', 'force', 'NodeLabel', {});
title("Graph for A_2");

%% Step 2-5
[clusters_1, L_1] = step2_5(A_1, k_1);
[clusters_2, L_2] = step2_5(A_2, k_2);

%% Step 6
% Assign different colors to each cluster
figure(3);
subplot(1,2,1);
h=plot(graph(A_1), 'Layout','force');
h.NodeCData = clusters_1;
% Display colorbar
title("Colored Graph A_1");
colorbar;
subplot(1,2,2);
h_2=plot(graph(A_2), 'Layout','force');
h_2.NodeCData = clusters_2;
% Display colorbar
title("Colored Graph A_2");
colorbar;

%% Sorted Fiedler Vector
[V_1,D] = eigs(L_1, 2, 'SA');
[V_2,D] = eigs(L_2, 2, 'SA');
figure(4);
subplot(1,2,1);
plot(sort(V_1(:,2)), '-');
title("Sorted Fiedler Vector A_1")
subplot(1,2,2);
plot(sort(V_2(:,2)), '-');
title("Sorted Fiedler Vector A_2");

%% Functions
function [A] = step1(E)
    col1 = E(:,1);
    col2 = E(:,2);
    max_ids = max(max(col1,col2));
    As= sparse(col1, col2, 1, max_ids, max_ids); 
    A = full(adjacency(graph(As)));
end
function [clusters, L] = step2_5(A, k)
    D = diag(sum(A,2));
    L = (D^(-1/2))*A*(D^(-1/2));
    
    %% Step 3
    [v, D] = eigs(L, k);
    
    %% Step 4
    Y = v./sum(v.*v,2).^(1/2);
    
    %% Step 5
    clusters = kmeans(Y, k);
end